pipeline/gcssink/sink.go

```diff
@@ -0,0 +1,123 @@
+	// If retrierConfig is not empty, the writer will be configured with the given retry options.
+	GetWriter(ctx context.Context, bucket, object, contentType string, chunkSize int, retrierConfig []storage.RetryOption) *storage.Writer
+	// Write writes the given message to the given writer.
+	Write(writer *storage.Writer, message SinkMessage) error
+	// Close closes the GCS client
+	Close() error
+}
+
+// gcsParallelWriter is a pipeline.Sink that writes messages to GCS in parallel.
+type gcsParallelWriter struct {
+	clientGateway GcsClientGateway
+	contentType   string
+	chunkSize     int
+	// retrierConfig is the configuration for the retrier. It is used to
+	// configure the retry police for the GCS client. Storage sdk provides
+	// some functions to create the retrier config. See:
+	// https://pkg.go.dev/cloud.google.com/go/storage#RetryOption
+	retrierConfig []storage.RetryOption
+}
+
+// SinkMessage is the input for the GCS Sink
+type SinkMessage struct {
+	Data        []byte
+	StoragePath string
+	Bucket      string
+	Metadata    map[string]string
+}
+
+// MustNew creates a new pipeline sink that write messages to GCS. It panics if
+// GcsClientGateway, contentType or chunkSize are nil or invalid. Order of the messages is not guaranteed.
+func MustNewParallel(
+	clientGateway GcsClientGateway,
+	contentType string,
+	chunkSize int,
+	retrierConfig []storage.RetryOption,
+) (pipeline.Sink, func() error) {
+	if clientGateway == nil {
+		panic("clientGateway is nil")
+	}
+
+	if contentType == "" {
+		panic("missing content type")
+	}
+
+	if chunkSize <= 0 {
+		panic("invalid chunk size")
+	}
+
+	writer := &gcsParallelWriter{
+		clientGateway: clientGateway,
+		contentType:   contentType,
+		chunkSize:     chunkSize,
+		retrierConfig: retrierConfig,
+	}
+
+	closeFn := func() error {
+		return clientGateway.Close()
+	}
+
+	return writer, closeFn
+}
+
+// Store implements the pipeline.Sink interface. It writes a batch of messages in parallel.
+// If any of the messages fail to be written, it returns an error. Order of the messages is not guaranteed.
+func (w *gcsParallelWriter) Store(ctx context.Context, messages ...pipeline.SinkMessage) error {
+	const op = errors.Op("gcssink.gcsParallelWriter.Store")
+
+	wg := &sync.WaitGroup{}
+
+	if len(messages) == 0 || messages == nil {
+		return nil
+	}
+
+	errs := make(chan error, len(messages))
+	for _, msg := range messages {
+		wg.Add(1)
+		go func(msg pipeline.SinkMessage) {
+			defer wg.Done()
+
+			sinkMsg, ok := msg.(SinkMessage)
+			if !ok {
+				errs <- errors.E(ErrInvalidSinkMessage, errors.KV("type", fmt.Sprintf("%T", msg)))
+				return
+			}
+
+			writer := w.clientGateway.GetWriter(ctx, sinkMsg.Bucket, sinkMsg.StoragePath, w.contentType, w.chunkSize, w.retrierConfig)
+			errs <- w.clientGateway.Write(writer, sinkMsg)
+
+		}(msg)
+	}
+
+	wg.Wait()
+
+	var sliceErrs []error
+	for range messages {
```

Comment:
Se for usar o errgroup, como comentado acima (recomendo que use), não precisaria considerar este comentário. 

Porém, essa abordagem é estranha. Imagino que foi feita para garantir que consumiu todos os erros. 

O correto seria primeiro criar novas rotinas para o produtor do canal de erros, ou seja, o range em cima do messages. O código criaria as rotinas e depois seguiria para o próximo trecho de código.

Esse próximo trecho deveria ser uma nova rotina que servirá para fechar o canal de erros logo após o `wg.Wait()`. É uma rotina usada exclusivamente para comunicar "olha, o wg.Wait acabou, terminou de processar todas as mensagens, pode fechar o canal de erros"

Por fim, precisaria ter um trecho que fica lendo do canal de erros, que deveria ser um range em cima do canal ao invés da forma que foi feita. Essa seria a rotina principal do método e ficará travada enquanto o canal de erros não for fechado, responsabilidade da rotina de comunicação, sendo que esta só irá fechar o cana de erros quando os produtores acabarem de consumir message, responsabilidade da rotina do produtor.

Então, em resumo, ficaria algo do tipo

```go
func Store() {
    errs := new(canal de erros)
    for _, msg := range messages {
        wg.Add(1)
        go func() {
            defer wg.Done()
            errs <- err
        }()
    }
    go func() {
        wg.Wait() // ficará travado aqui até acabar de processar tudo
        close(errs)
    }

    for err := range errs {
        // lidar com os erros
    }
}
```

Para entender um pouco mais: https://medium.com/engenharia-arquivei/consumindo-uma-lista-de-forma-eficiente-utilizando-go-d6ceb4efc63

Porém, o errgroup faz tudo isso por baixo, facilitando a nossa vida. Temos exemplos do uso dele em outros projetos